core/src/main/java/org/apache/calcite/util/ArrowSet.java

```diff
@@ -0,0 +1,357 @@
+ * limitations under the License.
+ */
+package org.apache.calcite.util;
+
+import com.google.common.collect.ImmutableSet;
+
+import java.util.ArrayDeque;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Represents a set of functional dependencies.
+ * Each functional dependency is represented as an {@link Arrow} from a set of
+ * determinant columns to a set of dependent columns.
+ *
+ * <p>An {@link ArrowSet} captures the complete set of functional dependencies
+ * that hold in a relation. Provides core algorithms in functional dependency theory.
+ *
+ * @see Arrow
+ * @see ImmutableBitSet
+ */
+public class ArrowSet {
+  public static final ArrowSet EMPTY = new ArrowSet(ImmutableSet.of());
+
+  // Maps each determinant set to the dependent set it functionally determines.
+  private final Map<ImmutableBitSet, ImmutableBitSet> dependencyGraph = new HashMap<>();
+
+  // Maps each column ordinal to its containing determinant sets for efficient reverse lookup.
+  private final Map<Integer, Set<ImmutableBitSet>> reverseIndex = new HashMap<>();
+
+  // All arrows in this ArrowSet.
+  private final Set<Arrow> arrowSet = new HashSet<>();
+
+  public ArrowSet(Set<Arrow> arrows) {
+    arrowSet.addAll(arrows);
+    for (Arrow arrow : arrows) {
+      ImmutableBitSet determinants = arrow.getDeterminants();
+      ImmutableBitSet dependents = arrow.getDependents();
+      dependencyGraph.merge(determinants, dependents, ImmutableBitSet::union);
+      for (int attr : determinants) {
+        reverseIndex.computeIfAbsent(attr, k -> new HashSet<>()).add(determinants);
+      }
+    }
+  }
+
+  public Set<ImmutableBitSet> getDeterminants(int attr) {
+    return reverseIndex.getOrDefault(attr, ImmutableSet.of());
+  }
+
+  public ImmutableBitSet getDependents(ImmutableBitSet determinants) {
+    return dependencyGraph.getOrDefault(determinants, ImmutableBitSet.of());
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  /**
+   * Computes the closure of a given set of column ordinals with respect to the current
+   * collection of functional dependencies.
+   *
+   * <p>The closure of a set of ordinals is defined as the set of all column ordinals
+   * that can be functionally determined from the specified set by applying zero or more
+   * functional dependencies present in this collection. This concept is fundamental
+   * in identifying keys, candidate keys, and in the normalization of relational schemas.
+   *
+   * <p>Example:
+   * <blockquote>
+   * <pre>
+   * // Given functional dependencies:
+   * {0} → {1}
+   * {1} → {2}
+   *
+   * // Closure computation:
+   * closure({0}) = {0, 1, 2}
+   * </pre>
+   * </blockquote>
+   *
+   * <p>Explanation:
```

Comment:
this is the implementation; this does not have to be in the JavaDoc, but in a normal comment.